src/commands/crashlytics-sourcemap-upload.ts

```diff
@@ -0,0 +1,178 @@
+import * as archiver from "archiver";
+import * as fs from "fs";
+import * as path from "path";
+import * as tmp from "tmp";
+import { statSync } from "fs-extra";
+import { readdirRecursive } from "../fsAsync";
+import { Command } from "../command";
+import { FirebaseError } from "../error";
+import { logLabeledBullet, logLabeledWarning } from "../utils";
+import { needProjectId } from "../projectUtils";
+import * as gcs from "../gcp/storage";
+import { getProjectNumber } from "../getProjectNumber";
+import { Options } from "../options";
+
+interface CommandOptions extends Options {
+  app?: string;
+  bucketLocation?: string;
+  appVersion?: string;
+}
+
+export const command = new Command("crashlytics:sourcemap:upload <mappingFiles>")
+  .description("upload javascript source maps to de-minify stack traces")
+  .option("--app <appID>", "the app id of your Firebase app")
+  .option(
+    "--bucket-location <bucketLocation>",
+    'the location of the Google Cloud Storage bucket (default: "US-CENTRAL1"',
+  )
+  .option(
+    "--app-version <appVersion>",
+    "the version of your Firebase app (defaults to Git commit hash, if available)",
+  )
+  .action(async (mappingFiles: string, options: CommandOptions) => {
+    checkGoogleAppID(options);
+
+    // App version
+    const appVersion = getAppVersion();
+
+    // Get project identifiers
+    const projectId = needProjectId(options);
+    const projectNumber = await getProjectNumber(options);
+
+    // Upsert default GCS bucket
+    const bucketName = await upsertBucket(projectId, projectNumber, options);
+
+    // Find and upload mapping files
+    const rootDir = options.projectRoot ?? process.cwd();
+    const filePath = path.relative(rootDir, mappingFiles);
+    let fstat: fs.Stats;
+    try {
+      fstat = statSync(filePath);
+    } catch (e) {
+      throw new FirebaseError(
+        "provide a valid file path or directory to mapping file(s), e.g. app/build/outputs/app.js.map or app/build/outputs",
+      );
+    }
+    if (fstat.isFile()) {
+      await uploadMap(mappingFiles, bucketName, appVersion, options);
+    } else if (fstat.isDirectory()) {
+      logLabeledBullet("crashlytics", "Looking for mapping files in your directory...");
+      const files = (
+        await readdirRecursive({ path: filePath, ignore: ["node_modules", ".git"] })
```

Comment:
I don't think the readdirRecursive function supports this yet, but reading files recursively with no maxDepth scares me (and has recently caused a incident for us). While you are here, could you add a maxDepth option to readdirRecursive, and set it to some reasonable value for this use case?